### PR TITLE
A 'cabal path' command.

### PR DESCRIPTION
**Pull request by athas**
Tools like [cabal2nix](https://github.com/NixOS/cabal2nix) need to access cabal-install's state or cache in order to do their job (in cabal2nix's case, to read the Hackage state).  Most of them didn't do this correctly in the first place, usually ignoring either `CABAL_CONFIG` or `CABAL_DIR`.  With the change to XDG, matters are even more dire, and it seems impractical for all tools to include the backwards compatibility logic.  One solution is for such tools to depend on `cabal-install` and import the logic from `Distribution.Client.Config`.  Of course, there is no guarantee that the version of `cabal-install` they are linked with is the same as the version of `cabal` that the user runs in their shell.

This draft PR adds a new command, `cabal dirs`, that prints out which directories `cabal` uses.  Example:

```
$ cabal dirs
cache-dir: /home/athas/.cache/cabal/packages
logs-dir: /home/athas/.cache/cabal/logs
store-dir: /home/athas/.local/state/cabal/store
```

These are not all the directories one could print.  For example, the various install-dirs are missing, mostly because I could not figure out how to actually obtain them whilst expanding any variables they contain.  (And what is "install-dirs global" used for anyway?)  However, the directories above are the ones that I have anecdotally observed that various third party tools are interested in.

An obvious alternative to this approach would be to extend `user-state` to also allow *reading* of arbitrary config values, but as far as I can see that is not possible: `user-state` has no idea what options are actually called, and simply uses the config file parser/unparser machinery.

---
Please include the following checklist in your PR:

* [X] Patches conform to the [coding conventions](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#conventions).
* [ ] Any changes that could be relevant to users [have been recorded in the changelog](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#changelog).
* [ ] The documentation has been updated, if necessary.

Please also shortly describe how you tested your change. Bonus points for added tests!

### CODE REVIEW

**Comment by fgaz** [comment]
> `user-state`

you mean `user-config`?

**Comment by athas** [comment]
Yes, I meant `user-config`.

**Comment by ulysses4ever** [comment]
@andreasabel requested something like this in https://github.com/haskell/cabal/issues/8855#issuecomment-1471042338

This sounds like a good idea. Maybe following `stack` convention would be wise (so, `cabal path`).

**Comment by andreasabel** [comment]
What is great about `stack path` is that you can get each directory individually as well.  This helps in automation and CI.
```
$ stack path --help
Usage: stack path [--stack-root] [--global-config] [--project-root] 
                  [--config-location] [--bin-path] [--programs] [--compiler-exe]
                  [--compiler-bin] [--compiler-tools-bin] [--local-bin] 
                  [--extra-include-dirs] [--extra-library-dirs] 
                  [--snapshot-pkg-db] [--local-pkg-db] [--global-pkg-db] 
                  [--ghc-package-path] [--snapshot-install-root] 
                  [--local-install-root] [--snapshot-doc-root] 
                  [--local-doc-root] [--local-hoogle-root] [--dist-dir] 
                  [--local-hpc-root] [--local-bin-path] [--ghc-paths] 
                  [--global-stack-root] [--setup-info-yaml URL] 
                  [--snapshot-location-base URL] [--help]

  Print out handy path information

Available options:
  --stack-root             Global Stack root directory
  --global-config          Global Stack configuration file
  --project-root           Project root (derived from stack.yaml file)
  --config-location        Configuration location (where the stack.yaml file is)
  --bin-path               PATH environment variable
  --programs               Install location for GHC and other core tools (see
                           'stack ls tools' command)
  --compiler-exe           Compiler binary (e.g. ghc)
  --compiler-bin           Directory containing the compiler binary (e.g. ghc)
  --compiler-tools-bin     Directory containing binaries specific to a
                           particular compiler (e.g. intero)
  --local-bin              Directory where Stack installs executables (e.g.
                           ~/.local/bin (Unix-like OSs) or %APPDATA%\local\bin
                           (Windows))
  --extra-include-dirs     Extra include directories
  --extra-library-dirs     Extra library directories
  --snapshot-pkg-db        Snapshot package database
  --local-pkg-db           Local project package database
  --global-pkg-db          Global package database
  --ghc-package-path       GHC_PACKAGE_PATH environment variable
  --snapshot-install-root  Snapshot installation root
  --local-install-root     Local project installation root
  --snapshot-doc-root      Snapshot documentation root
  --local-doc-root         Local project documentation root
  --local-hoogle-root      Local project documentation root
  --dist-dir               Dist work directory, relative to package directory
  --local-hpc-root         Where HPC reports and tix files are stored
  --local-bin-path         DEPRECATED: Use '--local-bin' instead
  --ghc-paths              DEPRECATED: Use '--programs' instead
  --global-stack-root      DEPRECATED: Use '--stack-root' instead
  --setup-info-yaml URL    Alternate URL or relative / absolute path for Stack
                           dependencies
  --snapshot-location-base URL
                           The base location of LTS/Nightly snapshots
```

**Comment by fgaz** [comment]
> An obvious alternative to this approach would be to extend [user-config] to also allow reading of arbitrary config values, but as far as I can see that is not possible: [user-config] has no idea what options are actually called, and simply uses the config file parser/unparser machinery.

I think it's possible, option names are in the `FieldDescr`s, See `configFieldDescriptions`.

**Comment by athas** [comment]
Apologies for the absence.  I have renamed the command to `cabal path` and rebased on top of master.

Can someone explain the format of `stack path`?  The help text above is not explicit, and for some reason e.g. `stack path --stack-root` just spits out some Nix error about installing GHC on my NixOS machine (not really clear to me why printing a path needs GHC, but whatever).

**Comment by athas** [comment]
I managed to find a system that could run `stack`.  The `cabal path` command now replicates its behaviour.

File: cabal-install/src/Distribution/Client/Main.hs [hunk 1]
```diff
@@ -1320,3 +1327,33 @@ manpageAction commands flags extraArgs _ = do
...
+  let getDir getDefault getGlobal =
+        maybe
+          getDefault
+          pure
+          (flagToMaybe $ getGlobal $ savedGlobalFlags cfg)
+      dirs =
+        [ ("cache-dir", getDir defaultCacheDir globalCacheDir)
+        , ("logs-dir", getDir defaultLogsDir globalLogsDir)
+        , ("store-dir", getDir defaultStoreDir globalStoreDir)
+        , ("config-file", getConfigFilePath (globalConfigFile globalFlags))
```
Comment (andreasabel) [inline cabal-install/src/Distribution/Client/Main.hs:1344]: I notice this list of open spellings (`something-dir`) is repeated below.
This isn't very resilient to accidental bug introductions, these lists can get out of sync (e.g. spelling mistakes).

Please do it the Haskell way:  Define an enum type with the 4 options and have a single place where you converts these into strings and back.

**Review by andreasabel — CHANGES_REQUESTED** [review]
Thanks @athas for this PR, very much appreciated.

I have a comment about using strings vs. using an enum type.

File: cabal-install/src/Distribution/Client/Main.hs (hunk 1 above)
Comment (athas) [reply cabal-install/src/Distribution/Client/Main.hs:1344]: The *best* way would be to have a list of *functions* that retrieve the corresponding paths inside the configuration record, but Cabal's convoluted module structure made that impossible without circular imports.  That frustrated me so much that I just put in the strings, but I suppose I should go back and add that sum type.

**Comment by fgaz** [comment]
@athas grep for `assertOutputContains` in `cabal-testsuite`

**Comment by athas** [comment]
Alright, I added some tests.  They do not test *exact* output, because the exact output depends on where the test suite is run from, as it includes file system paths.

**Comment by athas** [comment]
Wait, I can do better than that.

**Comment by athas** [comment]
Alright, these are tests that I am not ashamed of.

**Review by andreabedini — APPROVED** [review]

**Comment by andreabedini** [comment]
@andreasabel I take your suggestion about enums has been taken into consideration? Could you give another review?

**Review by andreasabel — APPROVED** [review]
Thanks, great work!

**Comment by athas** [comment]
I also just added a changelog entry.

**Comment by athas** [comment]
Does anything remain to be done here?

**Comment by ulysses4ever** [comment]
@athas thanks for the ping and your  patience. Rebasing…

**Comment by ulysses4ever** [comment]
The bot is behaving more and more weird over time. First, it stopped rebasing older PRs (since last week?). Now, it merged the PR before all the workflows finished...

File: cabal-install/src/Distribution/Client/Main.hs [hunk 2]
```diff
@@ -1347,3 +1359,32 @@ manpageAction commands flags extraArgs _ = do
           then dropExtension pname
           else pname
   manpageCmd cabalCmd commands flags
+
+pathAction :: PathFlags -> [String] -> Action
+pathAction pathflags extraArgs globalFlags = do
+  let verbosity = fromFlag (pathVerbosity pathflags)
+  unless (null extraArgs) $
+    dieWithException verbosity $
+      ManpageAction extraArgs
```
Comment (andreasabel) [inline cabal-install/src/Distribution/Client/Main.hs:1368]: This should be a new error `PathAction`, see issue:
- #9755

**Pull request merged**